q2_longitudinal/_utilities.py

```diff
@@ -683,9 +707,71 @@ def _nmit(table, sample_md, individual_id_column, corr_method="kendall",
     return _dist
 
 
+def _first_differences(metadata, state_column, individual_id_column, metric,
+                       replicate_handling='error', distance_matrix=None):
+
+    # let's force states to be numeric
+    _validate_is_numeric_column(metadata, state_column)
+
+    # create dummy group column in metadata so we can use downstream functions
+    # that split metadata by groups without actually bothering to do so.
+    group_column = "a_name_that_no_user_of_qiime2_could_conceivably_ever_use"
+    metadata[group_column] = 'null'
+
+    # calculate paired difference/distance distributions between each state
+    pairs_summary = pd.DataFrame()
+    errors = []
+    states = sorted(metadata[state_column].unique())
+    # iterate over range of sorted states in order to compare sequential states
+    for s in range(len(states) - 1):
+        # get pairs of samples at each sequential state
+        group_pairs, error = _get_group_pairs(
+            metadata, group_value='null',
+            individual_id_column=individual_id_column,
+            group_column=group_column, state_column=state_column,
+            state_values=[states[s], states[s + 1]],
+            replicate_handling=replicate_handling)
+        # compute distance between pairs
+        if distance_matrix is not None:
+            pairs, pairs_summaries = _extract_distance_distribution(
+                distance_matrix, group_pairs, metadata,
+                individual_id_column, group_column)
+        # or compute difference between pairs
+        else:
+            pairs, pairs_summaries = _get_pairwise_differences(
+                metadata, group_pairs, metric, individual_id_column,
+                group_column)
+        pairs_summary = pd.concat([pairs_summary, pairs_summaries])
+        errors.extend(error)
+
+    # convert pairs_summary to series with relevant metric
+    if distance_matrix is not None:
+        pairs_summary = pairs_summary['Distance']
+    else:
+        pairs_summary = pairs_summary['Difference']
+
+    # raise sensible error if output is empty.
+    if len(pairs_summary) == 0:
+        raise RuntimeError(
+            'Output is empty. Either no paired samples were detected in the '
+            'inputs or replicate samples were dropped. Check input files, '
+            'parameters, and replicate_handling settings.')
+
+    return pairs_summary
+
+
 def _validate_metadata_is_superset(metadata, table):
     metadata_ids = set(metadata.index.tolist())
     table_ids = set(table.index.tolist())
     if not table_ids.issubset(metadata_ids):
         raise ValueError('Missing samples in metadata: %r' %
                          table_ids.difference(metadata_ids))
+
+
+def _validate_is_numeric_column(metadata, metric):
```

Comment:
Can you have this utility also cast the column into a float dtype? I think it's possible for a user to supply a column that's composed of strings representing numbers -- I'm not sure what downstream code would do in that case with the "string numbers" (it might sort things lexicographically, which would be unexpected). Maybe worth a unit test too, something with states `1` and `10`?

Reply:
check out the new changes to `_add_metric_to_metadata`. If the metric is in metadata, it will already be cast to float by `_load_metadata`. The one metric not covered previously (I think) would have been metadata coming from feature tables.

See also new test `test_first_distances_numeric_values_represented_as_strings` to cover this.